README.md
[![NPM version](http://img.shields.io/npm/v/differential-browsers.svg)](https://www.npmjs.com/package/differential-browsers)

# differential-browsers

> 💡 This library is a fork of [Polymer/tools/packages/browser-capabilities](https://github.com/Polymer/tools/tree/master/packages/browser-capabilities), but the repository has been dead for two years. The biggest offender here is Anaheim edge that is now chromium based.


A JavaScript library that detects browser capabilities from a user agent string.

The following keywords are supported. See [differential-browsers.ts](https://github.com/zevisert/differential-browsers/blob/master/src/differential-browsers.ts) for the latest browser support matrix.

| Keyword       | Description
| :----         | :----
| push          | [HTTP/2 Server Push](https://developers.google.com/web/fundamentals/performance/http2/#server-push)
| serviceworker | [Service Worker API](https://developers.google.com/web/fundamentals/getting-started/primers/service-workers)
| modules       | [JavaScript Modules](https://www.chromestatus.com/feature/5365692190687232) (including dynamic `import()` and `import.meta`)
| es2015        | [ECMAScript 2015 (aka ES6)](https://developers.google.com/web/shows/ttt/series-2/es2015)
| es2016        | ECMAScript 2016
| es2017        | ECMAScript 2017
| es2018        | ECMAScript 2018

## TODO: 

- [ ] `push` is supported for `<link rel="preload" ...>`, but `modulepreload` is not available
- [ ] `es2019` - covering: `flat` and `flatMap` `Array` methods, try .. catch` optional binding, `Object.fromEntries`, stable `Array.sort`, and more
- [ ] `es2020` - covering: optional chaining (`?.`), nullish coalescing (`??`), `export * as namespace` and `BigInt`, and more


CHANGELOG.md
# Change Log

<!-- ## Unreleased -->
<!-- Add new, unreleased changes here. -->

## [1.2.0] - 2021-03-28
* Forked from Polymer/tools
* Updated for Anaheim Edge (Chromium Based)
* Updated Desktop Safari ES2018
  * Now passes the ES2018 predicate with AsyncIterator support but,
    (caution!) lacks regex lookbehind assertions
* Tests use Jest

## [1.1.4] - 2019-05-30
* Updated Firefox 67 support for modules (specifically because of dynamic import)

## [1.1.3] - 2018-11-15
* Updated Firefox 63 support for push.

## [1.1.2] - 2018-08-20
* Opera is now detected.

## [1.1.1] - 2018-06-28
* Chrome for iOS is now handled correctly, by delegating to the corresponding
  Safari feature matrix.
* Safari for iOS is handled more accurately, using iOS version.

## [1.1.0] - 2018-05-03
* Added new capabilities for es2016, es2017, and es2018.
* Dropped support for node v6. This is a soft break, as we aren't making any
  changes that are known to break node v6, but we're no longer testing against
  it. See our [node version support
  policy](https://www.polymer-project.org/2.0/docs/tools/node-support) for
  details.

## [1.0.0] - 2018-04-23
- Modules support now requires dynamic import and import.meta.
- Latest Safari now supports service workers and modules.

## [0.2.2] - 2018-01-01
- Add Chrome Headless browser.

## [0.2.1] - 2017-10-18
- Remove Firefox from set of browsers with push capability due to
  https://bugzilla.mozilla.org/show_bug.cgi?id=1409570.

## [0.2.0] - 2017-08-09
- Add JavaScript modules capability.

## [0.1.1] - 2017-08-08
- Declare TypeScript typings.

## [0.1.0] - 2017-08-08
- Initial release. Factored out of
  https://github.com/Polymer/prpl-server-node/blob/master/src/capabilities.ts


src/differential-browsers.test.ts
/**
 * @license
 * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import { expect } from "chai";
import { describe as suite, test } from "@jest/globals";
import * as capabilities from "./differential-browsers";

function assertBrowserCapabilities(
  userAgent: string,
  available: capabilities.BrowserCapability[]
) {
  expect([
    ...capabilities.browserCapabilities(userAgent),
  ]).to.include.all.members(available);
}

suite("capabilities", function () {
  test("unknown browser has no capabilities", () => {
    assertBrowserCapabilities("unknown browser", []);
  });

  test("chrome has all the capabilities", () => {
    assertBrowserCapabilities(
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_3) AppleWebKit/537.36" +
        " (KHTML, like Gecko) Chrome/66.0.3359.117 Safari/537.36",
      [
        "es2015",
        "es2016",
        "es2017",
        "es2018",
        "push",
        "serviceworker",
        "modules",
      ]
    );
  });

  test("chrome headless has all the capabilities", () => {
    assertBrowserCapabilities(
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_3) AppleWebKit/537.36" +
        " (KHTML, like Gecko) HeadlessChrome/66.0.3359.117 Safari/537.36",
      [
        "es2015",
        "es2016",
        "es2017",
        "es2018",
        "push",
        "serviceworker",
        "modules",
      ]
    );
  });

  test("edge es2015 support is predicated on minor browser version", () => {
    assertBrowserCapabilities(
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36" +
        " (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36 Edge/15.14986",
      ["push"]
    );

    assertBrowserCapabilities(
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36" +
        " (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36 Edge/15.15063",
      ["es2015", "es2016", "es2017", "push"]
    );
  });

  test("safari push capability is predicated on macOS version", () => {
    assertBrowserCapabilities(
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10) AppleWebKit/603.1.30" +
        " (KHTML, like Gecko) Version/10.1 Safari/603.1.30",
      ["es2015", "es2016", "es2017"]
    );

    assertBrowserCapabilities(
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11) AppleWebKit/603.1.30" +
        " (KHTML, like Gecko) Version/10.1 Safari/603.1.30",
      ["es2015", "es2016", "es2017", "push"]
    );
  });

  test("chrome on iOS is treated like safari mobile", () => {
    assertBrowserCapabilities(
      "Mozilla/5.0 (iPhone; CPU iPhone OS 9_3_3 like Mac OS X) AppleWebKit/601.1" +
        " (KHTML, like Gecko) CriOS/62.0.3202.70 Mobile/13G34 Safari/601.1.46 ",
      ["push"]
    );
    assertBrowserCapabilities(
      "Mozilla/5.0 (iPhone; CPU iPhone OS 10_3 like Mac OS X) AppleWebKit/602.1.50" +
        " (KHTML, like Gecko) CriOS/56.0.2924.75 Mobile/14E5239e Safari/602.1",
      ["es2015", "es2016", "es2017", "push"]
    );
    assertBrowserCapabilities(
      "Mozilla/5.0 (iPhone; CPU iPhone OS 11_3 like Mac OS X) AppleWebKit/604.1.34" +
        " (KHTML, like Gecko) CriOS/63.0.3239.73 Mobile/15E5167f Safari/604.1",
      ["es2015", "es2016", "es2017", "push", "serviceworker", "modules"]
    );
  });

  test("opera is detected", () => {
    assertBrowserCapabilities(
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36" +
        " (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36 OPR/52.0.2871.99",
      [
        "es2015",
        "es2016",
        "es2017",
        "es2018",
        "push",
        "serviceworker",
        "modules",
      ]
    );
  });

  test("parseVersion parses with fallback to -1", () => {
    expect(capabilities.parseVersion("37")).to.deep.equal([37]);
    expect(capabilities.parseVersion("37")).to.deep.equal([37]);
    expect(capabilities.parseVersion("10.987.00.1")).to.deep.equal([
      10,
      987,
      0,
      1,
    ]);
    expect(capabilities.parseVersion("4..foo.7")).to.deep.equal([4, -1, -1, 7]);
  });

  test("versionAtLeast checks all required parts", () => {
    expect(capabilities.versionAtLeast([3, 2, 1], [3, 2, 1])).to.be.true;
    expect(capabilities.versionAtLeast([3, 2, 1], [3, 2, 1, 4])).to.be.true;
    expect(capabilities.versionAtLeast([3, 2, 1], [4, 1, 0])).to.be.true;
    expect(capabilities.versionAtLeast([3, 2, 0], [3, 2])).to.be.true;

    expect(capabilities.versionAtLeast([3, 2, 1], [2, 2, 1])).to.be.false;
    expect(capabilities.versionAtLeast([3, 2, 1], [3, 1, 1])).to.be.false;
    expect(capabilities.versionAtLeast([3, 2, 1], [3, 1, 0])).to.be.false;
    expect(capabilities.versionAtLeast([3, 2, 1], [3, 2])).to.be.false;
    expect(capabilities.versionAtLeast([3, 2, 1], [3, 2])).to.be.false;
    expect(capabilities.versionAtLeast([3, 2, 1], [])).to.be.false;
  });
});


src/differential-browsers.ts
/**
 * @license
 * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import { UAParser } from "ua-parser-js";

/**
 * A feature supported by a web browser.
 */
export type BrowserCapability =
  // ECMAScript 2015 (aka ES6).
  | "es2015"
  // ECMAScript 2016.
  | "es2016"
  // ECMAScript 2017.
  | "es2017"
  // ECMAScript 2018.
  | "es2018"
  // HTTP/2 Server Push.
  | "push"
  // Service Worker API.
  | "serviceworker"
  // JavaScript modules, including dynamic import and import.meta.
  | "modules";

export type UserAgentPredicate = (ua: InstanceType<typeof UAParser>) => boolean;

const chrome = {
  es2015: since(49),
  es2016: since(58),
  es2017: since(58),
  es2018: since(64),
  push: since(41),
  serviceworker: since(45),
  modules: since(64),
};

const browserPredicates: {
  [browser: string]: { [key in BrowserCapability]: UserAgentPredicate };
} = {
  Chrome: chrome,
  Chromium: chrome,
  "Chrome Headless": chrome,
  Opera: {
    es2015: since(36),
    es2016: since(45),
    es2017: since(45),
    es2018: since(51),
    push: since(28),
    serviceworker: since(32),
    modules: since(48),
  },
  Vivaldi: {
    es2015: since(1),
    es2016: since(1, 14),
    es2017: since(1, 14),
    es2018: since(1, 14),
    push: since(1),
    serviceworker: since(1),
    modules: since(1, 14),
  },
  // Note that Safari intends to stop changing their user agent strings
  // (https://twitter.com/rmondello/status/943545865204989953). The details of
  // this are not yet clear, since recent versions do seem to be changing (at
  // least the OS bit). Be sure to actually test real user agents rather than
  // making assumptions based on release notes.
  "Mobile Safari": {
    es2015: sinceOS(10),
    es2016: sinceOS(10, 3),
    es2017: sinceOS(10, 3),
    es2018: () => false, // No async iterator or regex lookbehind assertions
    push: sinceOS(9, 2),
    serviceworker: sinceOS(11, 3),
    modules: sinceOS(11, 3),
  },
  Safari: {
    es2015: since(10),
    es2016: since(10, 1),
    es2017: since(10, 1),
    es2018: since(12), // Caution! No regex lookbehind assertions
    // HTTP/2 on desktop Safari requires macOS 10.11 according to
    // caniuse.com.
    push: (ua) => [since(9), sinceOS(10, 11)].every((pred) => pred(ua)),
    // https://webkit.org/status/#specification-service-workers
    serviceworker: since(11, 1),
    modules: since(11, 1),
  },
  Edge: {
    // Edge versions before 15.15063 may contain a JIT bug affecting ES6
    // constructors (https://github.com/Microsoft/ChakraCore/issues/1496).
    // Since this bug was fixed after es2016 and 2017 support, all these
    // versions are the same.
    es2015: since(15, 15063),
    es2016: since(15, 15063),
    es2017: since(15, 15063),
    es2018: since(79),
    push: since(12),
    // https://developer.microsoft.com/en-us/microsoft-edge/platform/status/serviceworker/
    serviceworker: since(79),
    modules: since(79),
  },
  Firefox: {
    es2015: since(51),
    es2016: since(52),
    es2017: since(52),
    es2018: since(58), // Except RegEx additions
    // Firefox bug - https://bugzilla.mozilla.org/show_bug.cgi?id=1409570
    push: since(63),
    serviceworker: since(44),
    modules: since(67),
  },
};

/**
 * Return the set of capabilities for a user agent string.
 */
export function browserCapabilities(userAgent: string): Set<BrowserCapability> {
  const ua = new UAParser(userAgent);
  let browserName = ua.getBrowser().name || "";
  if (browserName === "Chrome" && ua.getOS().name === "iOS") {
    // Chrome on iOS is really Safari.
    browserName = "Mobile Safari";
  }

  const predicates = browserPredicates[browserName] || {};
  return new Set(
    Array.from(
      Object.keys(predicates) as BrowserCapability[]
    ).filter((capability) => predicates[capability](ua))
  );
}

/**
 * Parse a "x.y.z" version string of any length into integer parts. Returns -1
 * for a part that doesn't parse.
 */
export function parseVersion(version: string | undefined): number[] {
  if (version == null) {
    return [];
  }
  return version.split(".").map((part) => {
    const i = parseInt(part, 10);
    return Number.isNaN(i) ? -1 : i;
  });
}

/**
 * Return whether `version` is at least as high as `atLeast`.
 */
export function versionAtLeast(atLeast: number[], version: number[]): boolean {
  for (let i = 0; i < atLeast.length; i++) {
    const r = atLeast[i];
    const v = version.length > i ? version[i] : 0;
    if (v > r) {
      return true;
    }
    if (v < r) {
      return false;
    }
  }
  return true;
}

/**
 * Make a predicate that checks if the browser version is at least this high.
 */
function since(...atLeast: number[]): UserAgentPredicate {
  return (ua) => versionAtLeast(atLeast, parseVersion(ua.getBrowser().version));
}

/**
 * Make a predicate that checks if the OS version is at least this high.
 */
function sinceOS(...atLeast: number[]): UserAgentPredicate {
  return (ua) => versionAtLeast(atLeast, parseVersion(ua.getOS().version));
}
